ibm/mas_devops/roles/ocp_setup_mas_deps/tasks/main.yml

```diff
@@ -28,13 +28,15 @@
 
 # TODO: Version check existing installation, ensure at least 1.2 is installed
 #       Fail if version < 1.2 is installed
-
 - name: Deploy Cert Manager
+  vars:
+    mas87_channels: "{{ mas87_default_channels }}"
   shell: |
     oc create namespace cert-manager
-    oc apply -f https://github.com/jetstack/cert-manager/releases/download/v1.2.0/cert-manager.yaml
+    oc apply -f https://github.com/jetstack/cert-manager/releases/download/v1.6.1/cert-manager.yaml
```

Comment:
where's the code that will install cert-manager 1.2 when MAS <8.7?  Am I missing something here, this looks like this will only work for MAS 8.7 + now, and we do nothing in the case of earlier versions ... we want the playbook to continue to install cert-manager 1.2 for MAS 8.6 (or less) installs as we don't want the hassle/don't have time to re-certify/re-test all prior releases against 1.6.

Reply:
True, good point - Added 1.6.1 as it's the supported by OCP 4.8, it seems then we'd need some logic to determine which version to install accordingly to the mas_channel as well?
i.e if not a MAS 8.7 then go with 1.2 version.

Problem is... if it is a MAS 8.7, we might have two options: with CPD 4.0 (i.e if using `lite-manage-roks.yml`) or without CPD 4.0 (i.e using `lite-core-roks.yml`

Any suggestions? 
I could simply rollback this to 1.2 if we assume MAS 8.7 will always use cert-manager that comes with CPD v4's...

Reply:
I can control which version of cert-manager to install based on ocp_version if that's ok?

If mas_channel is not a MAS 8.7, then jetstack cert-manager will be installed, but then we do a check based on the ocp_version, if this is an ocp 4.6 = install cm v1.2 otherwise install v1.6.1?
```
- name: Deploy Cert Manager
  vars:
    mas87_channels: "{{ mas87_default_channels }}"
    regex: '\d\.\d'
    os_version: "{{ ocp_version | regex_findall(regex) | join(',') }}"
    cm_version: "{{ 'v1.2' if (os_version == '4.6') else 'v1.6.1' }}" #if we have to install jetstack cert-manager, need to check what's the ocp version that is compatible
  shell: |
    oc create namespace cert-manager
    oc apply -f https://github.com/jetstack/cert-manager/releases/download/{{ cm_version }}/cert-manager.yaml
  when:
    - (_cert_manager_deployed.resources | length) == 0
    - mas_channel not in mas87_channels #Don't install this cert-manager if MAS 8.7 is being installed
    - not airgap_install
```

Reply:
if MAS 8.7 is set along with playbooks we still use mas_channel to determine the installation of cpd v4, and it will skip cert-manager standalone installation.

Reply:
We need to get on a call me, you and Joao to go over what myself and Joao have previously discussed about how cert-manager will be managed in 8.7 + 👍

Reply:
Added a piece of code to install foundation services + ibm-cert-manager operandRequest if we identify the need to install cert-manager in case of a MAS 8.7 channel being installed with the playbooks. 
Could you please review?

Reply:
Several possibilities :-)

Let me know if the following is aligned with what you guys have in mind:

* OCP 4.6 (MAS 8.6) + CP4D 3.5 (or no CP4D) -> cfg-ocp install cert-manager 1.2
* OCP 4.6 (MAS 8.6) + CP4D 4.0 -> cfg-ocp does nothing -> install-cp4d install cert-manager 1.6 (?) 
* OCP 4.8 (MAS 8.7) + no CP4D -> cfg-cop install cert-manager 1.6
* OCP 4.8 (MAS 8.7) + CP4D 4.0 -> cfg-ocp does nothing -> install-cp4d install cert-manager 1.6

Reply:
@alequint we had a call earlier today, and things got simpler than that. Basically `ocp_setup_mas_deps` role will install either foundation services ibm-cert-manager (which is the same one cpd v4 and MAS 8.7 will use) or will continue to install standalone/jetstack cert-manager if MAS channel **is not one** of the 8.7 channels (currently these are 'stable' and 'm2dev87'.) 
So in your 4 scenarios:
OCP 4.6 (MAS 8.6) + CP4D 3.5 (or no CP4D) -> cfg-ocp install cert-manager 1.2 - **MAS 8.6, so it will install cert-manager v1.2.0**
OCP 4.6 (MAS 8.6) + CP4D 4.0 -> cfg-ocp does nothing -> install-cp4d install cert-manager 1.6 (?) - **MAS 8.6, so it will install cert-manager v1.2.0, playbook will fail by the time it tries to install cpd v4 as MAS 8.6 is not compatible with it**
OCP 4.8 (MAS 8.7) + no CP4D -> cfg-cop install cert-manager 1.6 - **MAS 8.7, so it will install foundation services ibm-cert-manager**
OCP 4.8 (MAS 8.7) + CP4D 4.0 -> cfg-ocp does nothing -> install-cp4d install cert-manager 1.6 - **MAS 8.7, so it will install foundation services ibm-cert-manager, before cpd v4 do so which should be okay**

Reply:
@andrercm for the latest scenario, isn't there a chance to end up with two cert-managers (same version but duplicated installation - assuming they might be installed in different projects?). Or we are installing in the same project cp4d will do?

Reply:
@alequint don't think so because we'll request same ibm-cert-manager that foundation services provides, it's the same one that cpd v4 will request, we're just antecipating this step for this scenario. It will also be installed in ibm-common-services namespace.

Reply:
@andrercm here is a problem I noticed during my tests in this sequence of tasks:

```
- name: Check if Cert Manager is already installed
  community.kubernetes.k8s_info:
    api_version: v1
    name: cert-manager
    namespace: "cert-manager"
    kind: Deployment
  register: _cert_manager_deployed

- name: Define which Cert Manager to install
  set_fact:
    cm_version: "{{ 'foundation-services' if (mas_channel in mas87_channels) else 'standalone' }}" # install foundation services cert-manager if MAS 8.7+, otherwise, install jetstack/standalone version
  when:
    - (_cert_manager_deployed.resources | length) == 0
    - not airgap_install

- include_tasks: "tasks/certmanager/{{ cm_version }}.yml"
```

Let's say my cluster has cert-manager deployment already. This [criteria](https://github.com/ibm-mas/ansible-devops/blob/manage-components-cpd40/ibm/mas_devops/roles/ocp_setup_mas_deps/tasks/main.yml#L33) will fail, and so `cm_version` will not have any result. The following [task](https://github.com/ibm-mas/ansible-devops/blob/manage-components-cpd40/ibm/mas_devops/roles/ocp_setup_mas_deps/tasks/main.yml#L36), however, does not have same criteria, and it will have a runtime error.

Maybe it is just a matter of adding the criteria in the last task... however, there is not guarantee if the version already installed will support or not what automation is installing.



Is not used here as well

Reply:
ah yeah, you're right Alex, we can add same criteria in the include tasks in the bottom as safeguard in this case, just fixed it.
I think at this point we can just control whether it is installed or not, and if not then we install the compatible version with the "to-be" installed mas_channel.
If we need later to ensure cert-manager compatibility, then we would need to add more complexity to check the version and allow or not the mas install.

Reply:
Sounds like this is resolved, marking the conversation resolved.